core/src/main/scala/kafka/log/Log.scala

```diff
@@ -2686,3 +2670,50 @@ object LogMetricNames {
     List(NumLogSegments, LogStartOffset, LogEndOffset, Size)
   }
 }
+
+sealed trait SegmentDeletionReason {
+  def reasonString(log: Log, toDelete: Iterable[LogSegment]): String
+}
+
+case object RetentionMsBreachDeletion extends SegmentDeletionReason {
+  override def reasonString(log: Log, toDelete: Iterable[LogSegment]): String = {
+    s"Deleting segments due to retention time ${log.config.retentionMs}ms breach: ${toDelete.mkString(",")}"
+  }
+}
+
+case object RetentionSizeBreachDeletion extends SegmentDeletionReason {
+  override def reasonString(log: Log, toDelete: Iterable[LogSegment]): String = {
+    s"Deleting segments due to retention size ${log.config.retentionSize} breach. " +
+      s"Current log size is ${log.size}. ${toDelete.mkString(",")}"
```

Comment:
nit: replace "." with ":" 

`s"Current log size is ${log.size}: ${toDelete.mkString(",")}"`